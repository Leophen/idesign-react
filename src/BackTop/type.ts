export interface BackTopProps {
  /**
   * 类名
   */
  className?: string;
  /**
   * 自定义内容
   */
  children?: React.ReactNode;
  /**
   * 自定义样式
   */
  style?: React.CSSProperties;
  /**
   * 监听滚动的元素
   * @default () => window
   */
  target?: () => HTMLElement | Window;
  /**
   * 显示起点高度
   * @default 0
   */
  visibleHeight?: number
  /**
   * 是否平滑滚动
   * @default true
   */
  smooth?: boolean
  /**
   * 点击返回顶部时触发
   */
  onClick?: () => void;
  /**
   * 滚动层节点滚动时触发
   */
  onScroll?: () => void;
}
