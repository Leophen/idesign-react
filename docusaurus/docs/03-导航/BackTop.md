---
title: BackTop 回到顶部
---

返回页面顶部的操作按钮。

## 引入

```js
import {BackTop} from 'idesign';
```

## 基本用法

使用场景：

- 当页面内容区域比较长时；
- 当用户需要频繁返回顶部查看相关内容时。

```jsx live fffx
<>
    <BackTop visibilityHeight={0} />
    Scroll down to see the bottom-right
    <strong className="site-back-top-basic"> gray </strong>
    button.
</>
```

## API

:::caution 注意

有默认样式，距离底部 `50px`，可覆盖。

自定义样式宽高不大于 40px。

:::

| 参数             | 说明                                                          | 类型                | 默认值         |
| ---------------- | ------------------------------------------------------------- | ------------------- | -------------- |
| duration         | 回到顶部所需时间（ms）                                        | `number`            | `450`          |
| target           | 设置需要监听其滚动事件的元素，值为一个返回对应 DOM 元素的函数 | `() => HTMLElement` | `() => window` |
| visibilityHeight | 滚动高度达到此参数值才出现 BackTop                            | `number`            | `400`          |
| onClick          | 点击按钮的回调函数                                            | `function`          | `--`           |
