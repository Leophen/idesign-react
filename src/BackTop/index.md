---
nav:
  title: 组件
  path: /components
group:
  title: 导航组件
  order: 2
order: 7
---

# BackTop 回到顶部

可一键快捷返回顶部的按钮。

## 基本用法

对具体 `window` 滚动使用：

```tsx
import React from 'react';
import { BackTop } from 'idesign-react';

const App = () => {
  return (
    <BackTop />
  );
};

export default App;
```

## 具体节点滚动

对具体 `DOM` 节点滚动使用：

```tsx
import React from 'react';
import { BackTop } from 'idesign-react';

const App = () => {
  const data = Array(24).fill('item')
  return (
    <>
      <BackTop
        target={
          () => document.getElementById('idesign-demo-scroll1')
        }
      />
      <div id="idesign-demo-scroll1" style={{height: 400, overflow: 'auto'}}>
        {data.map((item, index) => (
          <li className="idesign-demo-itemy" key={index}>
            {item}
          </li>
        ))}
      </div>
    </>
  );
};

export default App;
```

## 滚动显示起点

通过 `visibleHeight` 属性控制滚动到指定高度时，才显示返回顶部按钮：

```tsx
import React from 'react';
import { BackTop } from 'idesign-react';

const App = () => {
  const data = Array(24).fill('item')
  return (
    <>
      <BackTop
        visibleHeight={100}
        target={
          () => document.getElementById('idesign-demo-scroll2')
        }
      />
      <div id="idesign-demo-scroll2" style={{height: 400, overflow: 'auto'}}>
        {data.map((item, index) => (
          <li className="idesign-demo-itemy" key={index}>
            {item}
          </li>
        ))}
      </div>
    </>
  );
};

export default App;
```

## 自定义返回顶部按钮

直接包裹自定义内容：

```tsx
import React from 'react';
import { BackTop } from 'idesign-react';

const App = () => {
  return (
    <BackTop>
      <div>⬆️</div>
    </BackTop>
  );
};

export default App;
```



<API />
