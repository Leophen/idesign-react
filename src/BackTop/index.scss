.i-back-top--button {
  width: 32px;
  padding: 0;
  .i-icon {
    color: #fff;
  }
}
