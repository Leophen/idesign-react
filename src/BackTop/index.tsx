import React, { useEffect, useRef, useState } from 'react';
import classNames from 'classnames';
import _ from 'lodash';
import './index.scss';
import Button from '../Button'
import Icon from '../Icon'
import Transition from '../Transition';
import { BackTopProps } from './type';

const BackTop: React.FC<BackTopProps> = (props) => {
  const {
    className,
    children,
    style,
    target = () => window,
    visibleHeight = 0,
    smooth = true,
    onClick = () => { },
    onScroll = () => { },
    ...restProps
  } = props;

  const [visible, setVisible] = useState(visibleHeight > 0 ? false : true);

  // 获取滚动层节点 t
  const getTarget = (target: HTMLElement | Window): HTMLElement => {
    return target === window ? document.documentElement : (target as HTMLElement);
  };

  const handleScroll = _.throttle((e: any) => {
    const top = (e.target as HTMLElement).scrollTop
    top >= visibleHeight ? setVisible(true) : setVisible(false)
    onScroll?.()
  }, 16)

  useEffect(() => {
    const t = getTarget(target?.())
    t?.addEventListener('scroll', handleScroll)
  }, [])

  const scrollToTop = () => {
    const t = getTarget(target?.())
    t?.scrollTo({
      top: 0,
      behavior: smooth ? 'smooth' : 'auto'
    });
    onClick?.();
  };

  return (
    <div
      className={classNames(
        'i-back-top',
        className
      )}
      style={{ ...style }}
      onClick={scrollToTop}
      {...restProps}
    >
      <Transition
        timeout={200}
        in={visible}
        animation='fade-in'
      >
        {children || (
          <Button shape="circle" className='i-back-top--button'>
            <Icon name="ArrowUpBold" />
          </Button>
        )}
      </Transition>
    </div>
  );
};

BackTop.displayName = 'BackTop';

export default BackTop;
