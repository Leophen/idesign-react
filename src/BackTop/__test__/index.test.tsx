import { render } from '@testing-library/react';
import BackTop from '../index';

describe('BackTop 组件测试', () => {
  it('init', () => {
    const { container } = render(<BackTop />);
    expect(container.firstChild).toHaveClass('i-back-top');
  })

  it('create', () => {
    const { asFragment } = render(<BackTop />);
    expect(asFragment()).toMatchSnapshot();
  });
});

