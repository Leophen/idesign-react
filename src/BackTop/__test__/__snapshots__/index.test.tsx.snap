// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BackTop 组件测试 create 1`] = `
<DocumentFragment>
  <div
    class="i-back-top"
  >
    <button
      class="i-button i-button--type-primary i-button--shape-circle i-back-top--button i-fade-in-appear i-fade-in-appear-active"
    >
      <i
        class="i-icon icon-ArrowUpBold"
      />
    </button>
  </div>
</DocumentFragment>
`;
